### ISSUE TITLE
enable multi core build

### ISSUE THREAD

**Issue by kdschlosser** [kiyolee/openssl1_1-win-build#6]
if you go into the project properties == >C/C++  ==> General 
there is an option in there you should turn on..
Multi-processor Compilation
The ding dongs at Microsoft did not name this correctly.

It should be "Say yes if you want your project to compile faster"
It  actually has not a thing to do with multiple processors kind of...

It launches cl multiple times.. cl can only operate on a single core. so by spawning multiple cl sessions each with a different file to compile you will then compile using 100% of your cpu..
The reason they have it termed like they do.. Technically speaking a core is a processor.. so 6 cores = 6 processors.. and because each time cl is spawned it will use a different processor (core) you then have the ability to use 100% of a multi core processor. That setting determines the number of cl sessions to spawn that matches the number of processors the system has.
8 processors  (cores) = 8 cl processes that get spawned..

![image](https://user-images.githubusercontent.com/10932297/71789735-19355600-2fea-11ea-893f-f615d08a70c9.png)

This should be turned on by default I would imagine.. I do not know of anyone that has a single core machine anymore.. and if they do it would be a recommendation to not compile OpenSSL on it

**Comment by kiyolee** [comment]
I know about the option but I don't usually put that into any project.
As you probably now find out, the option has a nasty effect if you have also enabled "parrallel project builds" through "Build and Run" settings.
VS has too modes of parallel builds:
1. The original "Build and Run" setting simply start multiple cl.exe and compile sources sequentially per cl.exe. No parallelism within cl.exe.
2. The "Multi-processor Compilation" as shown, which is relatively newer, that runs cl.exe in multithreading mode that compiles multiple sources in parallel. The idea was that multiple threads inside the same cl.exe can share precompiled information in memory and hence speeds things up.
If you enable both options, say for 8 cores, you end up having 8x8 = 64 threads that are all compiling sources. That usually becomes overloading a system and just being counter effective.
When I am in IDE, I usually only use "Build and Run". To really want to build things fast, I will run msbuild with option /m at the command line then it will effectively using "Multi-processor Compilation". I leave the option out of projects is that I don't want to require people adjusting the "Build and Run" option.
If you do want to use "Multi-process Compilation" inside IDE, you need to reduce the "Build and Run" setting to prevent overloading.

**Comment by kdschlosser** [comment]
OK you have more of a handle on this then I have.. I do not know the inner mechanics of how the thing works i just know that if it is not enabled it is slower then hell. LOL

So now I am going to hit you with yet another question regarding this..

which is faster serial project compiling with parallel compilers
or parallel projects with serial compilers??
 
if using a solution that has 8 projects and each of the 8 projects have 50 files to be compiled that there shouldn't be to much difference between them..

But in the case of this solution where it is 4 projects if using serial project builds on an 8 core machine you would only be able to use 1/2 of the CPU.

the other spin on this would be what happens when one project has 300 files say and another has 3.. and the one that has 3 finishes. the one that has 300 is still going to hokey poke along using only one core. 

My thought process on this is probably incorrect.. but it sounds like running multiple compiler in order to max the CPU is going to provide the best possible speed increase no matter the number of projects, number of files or number of CPU cores.

**Comment by kiyolee** [comment]
I have never done a thorough building speed comparison.
But I do feel that building with `msbuild /m` seems to be faster than building inside IDE.
The GUI update in IDE may slow things down a little bit as well I suppose. Can't truly tell.
That can also be noticed through the CPU usage meter.
`msbuild /m` seems to saturate the CPU cores more offen.
Unfortunately, msbuild does not has an option to build like VS IDE.
I should try building with devenv.exe one day which should be more close to the speed of inside IDE.
Btw, VS may also cache various different things in various different states at various different times, building time in VS IDE is not always easily being consistent.

**Comment by kiyolee** [comment]
Forget to mention projects dependencies also play a role here.
Projects depending on each other are to be build sequentially.
This is not simple matter.

**Comment by kdschlosser** [comment]
I have written a few build systems that use cl to compile c code into python extensions. I have also used msbuild to build the same extension using a solution..

what I do when calling cl is i set up n + 1 threads n = number fo cpu cores

i take the list of source files and split it up into n equal parts.. I pass each part to each thread. any output gets written to a queue and the main thread is responsible for writing the output to the screen. that removes the need to use threads loke to have the output be correct.

once the 7 threads terminate the main thread then calls link..

the time to compile using msbuild was in the 3 minute area.. 
the time to build using the python script was a tad over 30 seconds.

**Comment by kdschlosser** [comment]
I am bringing this up because I need to shave down that compile time to as small as possible.. 
Using a CI for free typically has a time limit of 60 minutes.. when compiling an extension for python it is suggested to compile the extension with the same compiler version that was used to compile python.. to cover 4 python versions i need to be able to compile with 3 different msvc version.. and I have to do this for each architecture so a total of 6 compilations... Trying to squeeze that in about 50 minutes is going to be hard to do... need to be at the 50 minute mark and not the 60. need some wiggle room because of CI server load..

I have never even checked to see how many cores are supplied with a CI seerver VM. if it is only a single vcore then this whole thing is going to be moot anyway

**Comment by kdschlosser** [comment]
i just checked and they supply 2 vcores

**Comment by kiyolee** [comment]
Interesting, hope my solution helps.
